scylla/src/frame/value.rs

```diff
@@ -1,171 +1,435 @@
-        match self {
-            Some(val) => <T as Into<Value>>::into(val),
-            None => Value::Null,
-        }
+//
+//  Value impls
+//
+
+// Implement Value for primitive types
+impl Value for i8 {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        buf.put_i32(1);
+        buf.put_i8(*self);
+        Ok(())
+    }
+}
+
+impl Value for i16 {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        buf.put_i32(2);
+        buf.put_i16(*self);
+        Ok(())
+    }
+}
+
+impl Value for i32 {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        buf.put_i32(4);
+        buf.put_i32(*self);
+        Ok(())
+    }
+}
+
+impl Value for i64 {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        buf.put_i32(8);
+        buf.put_i64(*self);
+        Ok(())
+    }
+}
+
+impl Value for &str {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        let str_bytes: &[u8] = self.as_bytes();
+        let val_len: i32 = str_bytes.len().try_into().map_err(|_| ValueTooBig)?;
+
+        buf.put_i32(val_len);
+        buf.put(str_bytes);
+
+        Ok(())
     }
 }
 
-// Given a type that implements Into<Value> implement TryIntoValue for it
-// trying to do that with generics fails because TryIntoValue<Option<_>> is ambiguous
-macro_rules! impl_try_into_for_into {
-    ($t:ty) => {
-        impl TryIntoValue for $t {
-            fn try_into_value(self) -> Result<Value, ValueTooBig> {
-                Ok(self.into())
+impl Value for String {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        <&str as Value>::serialize(&self.as_str(), buf)
+    }
+}
+
+/// Every Option<T> can be serialized as None -> NULL, Some(val) -> val.serialize()
+impl<T: Value> Value for Option<T> {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        match self {
+            Some(val) => <T as Value>::serialize(val, buf),
+            None => {
+                buf.put_i32(-1);
+                Ok(())
             }
         }
-    };
+    }
 }
 
-impl Into<Value> for String {
-    fn into(self) -> Value {
-        Value::Val(self.into())
+// Every &impl Value should also implement Value
+impl<T: Value> Value for &T {
+    fn serialize(&self, buf: &mut Vec<u8>) -> Result<(), ValueTooBig> {
+        <T as Value>::serialize(*self, buf)
     }
 }
 
-impl_try_into_for_into!(String);
+//
+//  ValueList impls
+//
 
-impl Into<Value> for &str {
-    fn into(self) -> Value {
-        Value::Val(self.to_owned().into())
+// Implement ValueList for the unit type
+impl ValueList for () {
+    fn serialized(&self) -> SerializedResult<'_> {
+        Ok(Cow::Owned(SerializedValues::new()))
     }
 }
 
-impl_try_into_for_into!(&str);
-
-impl Into<Value> for i8 {
-    fn into(self) -> Value {
-        Value::Val(self.to_be_bytes().to_vec().into())
+impl ValueList for [u8; 0] {
```

Comment:
It is used to allow passing &[] to query as values argument. It's only for u8 because otherwise Rust complains that it is not able to infer type. I should've added a comment - gonna do that.